components/roster/heroes/Heroes.tsx

```diff
@@ -0,0 +1,283 @@
+  label: string;
+  disabled: boolean;
+};
+
+type HeroProps = {
+  heroList: Hero[];
+};
+
+const isAscensionLevelValid = (level: AscensionLevel | null): boolean => {
+  if (!level) return false;
+  const validLevels = Object.keys(AscensionLevel).splice(7);
+  return validLevels.includes(level);
+};
+
+const exclusiveEquipmentOptions = Array.from(
+  { length: 6 },
+  (_, i) => `${i * 5}`,
+);
+
+type HeroChange = {
+  property: "unlock" | "ascension" | "equipment";
+  value: string;
+  hero: Hero;
+};
+
+function reducer(heroes: Hero[], change: HeroChange) {
+  const property = change.property;
+  const value = change.value;
+  const hero = change.hero;
+
+  switch (property) {
+    case "unlock": {
+      return heroes.map((h) =>
+        h.key === hero.key
+          ? { ...h, unlocked: JSON.parse(value.toLowerCase()) }
+          : h,
+      );
+    }
+    case "ascension": {
+      let ascensionLevel = (
+        Object.keys(AscensionLevel) as Array<keyof typeof AscensionLevel>
+      ).find((key) => AscensionLevel[key] === value) as AscensionLevel;
+
+      if (!isAscensionLevelValid(ascensionLevel)) {
+        return heroes.map((h) =>
+          h.key === hero.key
+            ? { ...h, ascension: ascensionLevel, exEquipment: 0 }
+            : h,
+        );
+      }
+
+      return heroes.map((h) =>
+        h.key === hero.key ? { ...h, ascension: ascensionLevel } : h,
+      );
+    }
+    case "equipment": {
+      return heroes.map((h) =>
+        h.key === hero.key ? { ...h, exEquipment: parseInt(value) } : h,
+      );
+    }
+    default:
+      return heroes;
+  }
+}
+
+export default function Heroes({ heroList }: HeroProps) {
+  const [heroes, dispatch] = useReducer(reducer, heroList);
+
+  const CustomSelectSubmenu = ({
+    hero,
+    options,
+    selected,
+    heroProperty,
+    label,
+    disabled = false,
+  }: SubSelectProps) => (
+    <DropdownMenuSub>
+      <DropdownMenuSubTrigger
+        className={`w-full ${disabled ? "opacity-50 cursor-not-allowed" : ""}`}
+        disabled={disabled}
+      >
+        <span>{label}</span>
+      </DropdownMenuSubTrigger>
+      {!disabled && (
+        <DropdownMenuSubContent className="w-48">
+          <DropdownMenuLabel>{label}</DropdownMenuLabel>
+          <DropdownMenuSeparator />
+
+          <DropdownMenuSeparator />
+          {options.map((option) => (
+            <DropdownMenuItem
+              key={option}
+              onSelect={() =>
+                dispatch({ property: heroProperty, value: option, hero: hero })
+              }
+            >
+              <Check
+                className={`mr-2 h-4 w-4 ${
+                  selected === option ? "opacity-100" : "opacity-0"
+                }`}
+              />
+              {option}
+            </DropdownMenuItem>
+          ))}
+        </DropdownMenuSubContent>
+      )}
+    </DropdownMenuSub>
+  );
+
+  const handleSave = async (e: FormEvent) => {
+    e.preventDefault();
+
+    let items = heroes
```

Comment:
this is not a `let`, also see previous comment about whether this is necessary or not. i think the api endpoint should be able to handle the data stored in the state variable properly